src/audio/asrc/asrc_farrow.c

```diff
@@ -551,52 +539,51 @@ enum asrc_error_code asrc_set_output_format(struct comp_dev *dev,
 /*
  * BUFFER FUNCTIONS
  */
-static enum asrc_error_code initialise_buffer(struct asrc_farrow *src_obj)
+static enum asrc_error_code initialise_buffer(struct comp_dev *dev,
+					      struct asrc_farrow *src_obj)
 {
-	uint8_t *buffer;
+	int32_t *start_32;
+	int16_t *start_16;
 	int ch;
-	int n;
-
-	/*
-	 * base_address points to the first address subsequently to the
-	 * memory where the pointers to each ring buffer are stored.
-	 */
-	if (src_obj->bit_depth == 32)
-		buffer = (uint8_t *)(src_obj->ring_buffers32 +
-			src_obj->num_channels);
-	else if (src_obj->bit_depth == 16)
-		buffer = (uint8_t *)(src_obj->ring_buffers16 +
-			src_obj->num_channels);
 
 	/*
-	 * set buffer_length to filter_length * 2 to compensate for
+	 * Set buffer_length to filter_length * 2 to compensate for
 	 * missing element wise wrap around while loading but allowing
 	 * aligned loads.
 	 */
 	src_obj->buffer_length = src_obj->filter_length * 2;
+	if (src_obj->buffer_length > ASRC_MAX_BUFFER_LENGTH) {
+		comp_err(dev, "initialise_buffer(), buffer_length %d exceeds max.",
+			 src_obj->buffer_length);
+		return ASRC_EC_INVALID_BUFFER_LENGTH;
+	}
+
 	src_obj->buffer_write_position = src_obj->filter_length;
 
-	/* set the base addresses for every channel and initialise the
-	 * buffers to zero
+	/*
+	 * Initialize the dynamically allocated 2D array and clear the
+	 * buffers to zero.
 	 */
 	if (src_obj->bit_depth == 32) {
-		for (ch = 0; ch < src_obj->num_channels; ch++) {
-			src_obj->ring_buffers32[ch] = ((int32_t *)buffer) +
+		start_32 = (int32_t *)(src_obj->ring_buffers32 +
+			src_obj->num_channels);
+		for (ch = 0; ch < src_obj->num_channels; ch++)
+			src_obj->ring_buffers32[ch] = start_32 +
 				ch * src_obj->buffer_length;
 
-			/* initialise to zero */
-			for (n = 0; n < src_obj->buffer_length; n++)
-				src_obj->ring_buffers32[ch][n] = 0;
-		}
-	} else if (src_obj->bit_depth == 16) {
-		for (ch = 0; ch < src_obj->num_channels; ch++) {
-			src_obj->ring_buffers16[ch] = ((int16_t *)buffer) +
+		/* initialise to zero */
+		memset(start_32, 0, src_obj->num_channels *
+			src_obj->buffer_length * sizeof(int32_t));
+	} else {
+		start_16 = (int16_t *)(src_obj->ring_buffers16 +
+			src_obj->num_channels);
+		for (ch = 0; ch < src_obj->num_channels; ch++)
+			src_obj->ring_buffers16[ch] = start_16 +
 				ch * src_obj->buffer_length;
 
-			/* initialise to zero */
-			for (n = 0; n < src_obj->buffer_length; n++)
-				src_obj->ring_buffers16[ch][n] = 0;
-		}
+		/* initialise to zero */
+		memset(start_16, 0, src_obj->num_channels *
+			src_obj->buffer_length * sizeof(int16_t));
 	}
```

Comment:
Today I learned something new about C... And it's even in K&R and I wasn't aware of it. Here's the thing: `&(a[1]) - &(a[0]) == 1` for both `int a[2];` and `int *a;` and in both cases it's known at compile-time. However `&(a[0][1]) - &(a[0][0])` is also valid for both `int a[2][3];` and e.g. `int *a[2];` but - for the former it's also known at compile time and is always 1 whereas for the latter it depends on values of `a[0]` and `a[1]` and consequently only can be calculated at run-time... Maybe it's obvious to everyone here except me, then we can certainly use this code, but if I'm not the only one who finds it... confusing - to put it mildly - maybe we want to make the code a bit clearer about what exactly we mean :-)

Reply:
@lyakh How does it depend on a[1] when both values are read from a[0]?

Reply:
@paulstelian97 yes, sorry, see - it's that confusing (for me) :-) Make that `&(a[1][0]) - &(a[0][0]);`

Reply:
For that one yeah, it is statically known for 2D arrays but determined as an actual runtime computation for jagged arrays (double pointer).

Reply:
Please share your proposal how to make the code a bit clearer.

(I'd like to do a rewrite the data I/O part as this wastes some RAM but I'll do after I've managed to fix the stress test glitches issue. I would remove the "2D" array and use normal channel interleaved circular buffers for FIR as I use in other components like EQ and SRC).

Reply:
@singalsu as long as you use a dynamically allocated 2D array like `int **a;` or `int *a[];` and you refer to individual elements using `a[i][j]` there doesn't seem to be much choice - C wants to have an array of pointers exactly like it's done now. If you can do that as an interleaved array and it would be easier to work with - that could be good, but if you're also fine with the current implementation - it's correct too, basically it's just me who wasn't aware of this C... peculiarity...

Reply:
I'd like get this simple fix merged unless there are still violations against good practices. I don't promise when but I will do the rewrite.

If it would be just for MCPS improve it would be low priority (current speed should be quite acceptable) but the RAM usage is much more than needed. It's not a large task after there's a stable version that passes the testing. I have now fix PR for 2 of 3 different type know bug. The third is under work. If that fixes remaining issues the rewrite can be done.